File: Neetcode-150/BinarySearch/FindMinimumInRotatedSortedArray.py
# Time Complexity: O(logn)
# Space Complexity: O(1)
'''
Wow its like midnight rn and I still kinda got the soln to the problem. I saw this problem a long time ago so I remember remnants of the gist of the soln.
Though yeah, the time complexity is O(logn) coz we're using binary search, and the space complexity is O(1) coz we're not using any extra data structures. 
'''

def findMin(nums):

    left, right = 0, len(nums) - 1

    while(left <= right):

        mid = left + (right - left)//2

        if(nums[right] < nums[mid]):
            left = mid + 1
        elif(nums[right] >= nums[mid]):
            right = mid - 1

        if(mid-1 > 0 and nums[mid-1] > nums[mid]):
            return nums[mid]
        elif(mid+1 < len(nums) and nums[mid+1] < nums[mid]):
            return nums[mid+1]
        elif(mid == 0):
            # This took me a while, if left and right are the same, then we've
            # reached the target
            return nums[mid]
        

File: Neetcode-150/BinarySearch/find_minimum_in_rotated_sorted_array.java
public class find_minimum_in_rotated_sorted_array {
    public static int findMin(int[] nums) {
        /*
        Brute Force O(n) approach

        int min = nums[0];
        for (int n : nums) {
            min = Math.min(min, n);
        }
        return min;
        */

        int left = 0, right = nums.length - 1, mid; // pointers for binary search

        while (left <= right) { // applying binary search approach, as long as left and right dont cross
            mid = left + (right - left) / 2; // calculate mid

            if (mid > 0 && nums[mid] < nums[mid - 1]) { // since we know its sorted and it can only be rotated, find the part that isnt sorted because if the number before mid is greater than this portion is not sorted and its where the rotation occured, meaning this is the minimum
                return nums[mid];
            }
            else if (nums[left] <= nums[mid] && nums[mid] > nums[right]) { // otherwise if the left side is sorted shift to try and find the unsorted portion
                left = mid + 1;
            }
            else { // apply same here and shift right pointer
                right = mid - 1;
            }
        }

        return nums[left];
    }
    public static void main(String[] args) {
        int[] nums = {3,4,5,6,1,2}, nums2 = {4,5,0,1,2,3};

        System.out.println(findMin(nums));
        System.out.println(findMin(nums2));

        // time complexity of this is O(log(n)) because of binary search and the space complexity is O(1)
    }
}
